File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        Map<String,String> mp = new HashMap<>();
        for(List<String> el : knowledge){
            mp.put(el.get(0),el.get(1));
        }
        StringBuilder res = new StringBuilder();
        int st = -1;
        for(int i = 0 ; i < s.length(); i++){
            char c = s.charAt(i);
            if(c == '('){
                st = i;

            }
            else if(c == ')'){
                String k = s.substring(st + 1 , i);
                res.append(mp.getOrDefault(k,"?"));
                st = -1;
            }
            else if(st < 0){
                res.append(c);
            }
        }
        return res.toString();

    }
}

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        map<string,string> mp;
        for(vector<string> el : knowledge){
            mp[el[0]] = el[1];
        }
        string res = "";
        int st = -1;
        for(int i = 0 ; i < s.length() ; i++){
            char c = s[i];
            if(c == '('){
                st =  i;
            }
            else if (c == ')'){
                string k = s.substr(st+1 , i - st - 1);
                if(mp.count(k)){
                 res += mp[k];
                }
                else{
                res += "?";
                }
                st = -1;
            }

            else if(st < 0){
                res += s[i];
            }
        }
        return res;
    }
};